[미션] 검색자동완성

dev-angelo opened the pull request:
Demo page: https://dev-angelo.github.io/amazon-clone/

**완료 사항 >**

* 검색어를 입력하면 자동완성결과가 노출된다. 
* 자동완성결과는 바로 추가되지 않고, 300ms지연 후에 화면에 추가된다. 
* 입력창의 내용을 백스페이스로 삭제해도 일치하는 자동완성결과가 노출된다. 
* 자동완성 결과는 키보드 위/아래키로 이동할수 있다.
* 자동완성 결과를 키보드 방향키로 이동시에 선택부분의 배경색은 변경된다. 선택된 상태에서 엔터키를 입력하면 해당검색어가 위쪽 검색input창에 추가된다.  동시에 검색결과창은 사라진다.
* 검색버튼을 눌러도 검색이 이뤄지진 않으며, 자동완성 결과 창은 닫힌다. 

**미완료 사항 >**
* 노출된 데이터 중 검색어와 일치하는 단어는 색깔이 하이라이트 되여 보여진다.
* (선택기능) 검색창에 포커스가 가면, 최근 검색한 결과가 최대 5개까지 노출된다. (시간 역순으로 최근 검색한 내용이 위에 나옴)

**힘들었던 부분 >**
* MVC 패턴 을 이 익숙하지 않아서 (기존의 지식은 특정 프레임워크에 특화된 
  MVC 패턴? 이었습니다.) 구조가 헷갈려 코딩을 하면서도 의문이 들었습니다.

**아쉬웠던 부분 >**
* Browser 에서는 EventEmitter 를 사용 할 수 없어서 많은 부분을 callback 으로 처리.

**To do >**
 **요구사항 관련:**
 * 노출된 데이터 중 검색어와 일치하는 단어는 색깔이 하이라이트 되여 보여진다.
 * (선택기능) 검색창에 포커스가 가면, 최근 검색한 결과가 최대 5개까지 노출된다. 
   (시간 역순으로 최근 검색한 내용이 위에 나옴)

**기타:**
  * MVC 패턴을 더 공부하여 구조 변경 필요시 구조 변경.
  * 기 구현되어 있던 Slide 를 MVC 패턴으로 마이그레이션.

**그 외 >**
* Demo 를 위한 Github Pages 는 http 프로토콜을 허용하지 않아서 의도하지는 않았지만
  Apache Server, DDNS, SSL, SSL 인증키 발급 방법, SSL 인증키를 express 서버에 적용 등을
  알게 되었습니다. 

**질문사항 >**
* View 는 사용자에게 보여지는 내용들을 위한 계층인데, 
  render 함수 외에도 사용자에게 화면이 변경되는 내용들을 설정하는 DOM 조작 

  1. controller 가 서버로부터 받은 response 로 model 을 설정 하고,
  2. controller 는 model 로부터에서 suggestion list 가 변경되었다는 것을 전달 받아 
  3. 그 내용을 view 로 전달하여
  4. view 에서는 appendChild 등의 함수를 통해 사용자에게 보여질 자동완성 리스트 구성 등

  도 View 에서의 역할이라고 생각 할 수 있을까요?

dev-angelo committed:
[수정] Slide 관련 파일들을 Slide 폴더로 이동

Slide 관련 파일들을 Slide 폴더로 이동.

dev-angelo committed:
[수정] SlideService.js 참조 경로 수정

SlideService.js 참조 경로 수정.

Why >
Slide 관련 js 파일들이 Slide 폴더로 이동되었기 때문.

dev-angelo committed:
[수정] 자동완성 기능을 위한 html, css 수정

자동완성 기능을 위한 input field, background, list 등을 추가.

dev-angelo committed:
[파일추가 및 수정] post 요청 처리 기능 추가

post 요청 처리 기능 추가.

Server.js >
* Client 의 post 요청 (사용자 입력에 대한 suggestion list 반환)
  을 처리한는 기능 추가.

  -> words 배열에서 사용자 입력을 시작으로 하는 단어 배열을
     Client 에게 response 로 전달

words.js (New file) >
* Alphabet 순서로 정렬된 영단어 배열 추가

dev-angelo committed:
[파일추가] Search 기능을 위한 Skeleton Class/Code 추가

Search 기능을 위한 Skeleton Class/Code 추가.

dev-angelo committed:
[수정] 객체 생성 시 rendering 을 하도록 수정

html 파일에 정의되어 있던 element 들을 View 에서 렌더링 하도록 수정.

dev-angelo committed:
[수정] CSS 파일 수정

CSS 파일 수정

style.css >
1. 불필요한 라인 제거.
2. string 을 사용한 color 값을 rgb 로 수정.

dev-angelo committed:
[수정] Server 가 response 하는 array 의 갯수를 최대 10개로 수정

Client 의 suggestion 요청에 대한 array 를 전달 할 때,
array 의 요소 갯수가 최대 10개가 되도로 수정.

Why >
1. Amazon 에서도 최대 10 개의 list 를 보여주는것을 확인.
2. Front End 레벨에서 10 개를 보여준다는 requirement 가 있다면,
   서버에서 불필요한 array 구성을 위한 시간을 줄일 수 있고
   array 크기가 줄어들어서 적은 네트워크 사용으로 빠른 전달이 가능하고
   Front End 에서도 불필요한 로직이 필요 없어지게 됨.

dev-angelo committed:
[수정] 기존 Slide 에서의 데이터 요청 url 수정

기존 Slide 에서의 데이터 요청 url 수정.

Why >
* 포트포워딩을 통해 express 서버에서 외부 접속이 가능해졌기 때문.

dev-angelo committed:
[수정] 사용자 입력에 대한 Autocompletion 출력 기능 추가

사용자 입력에 대한 Autocompletion 출력 기능 추가.

Why >
* 요구사항

구현사항 >
1. 검색어 입력 시 자동완성결과 노출되돌 구현.
   (결과가 있다면)

2. 검색어 입력 시 dim 처리 된 background 가 노출되도록 구현
   (결과가 있다면)

3. 자동완성 list 에서 단어 선택 시,
   3-1. input field 에 추가되고
   3-2. 자동완성 결과를 사라지고
   3-3. dim 처리 된 background 를 사라지도록 구현.

4. 자동완성 list 가 보여지고 있을 때 dim 처리 된 background 클릭 시,
   4-1. 자동완성 결과가 사라지고
   4-2. dim 처리 된 background 가 사라지도록 구현.

5. 자동완성 결과가 나타나고 있는 상태에서 백스페이스로 삭제해도
   일치하는 자동완성결과가 노출되도록 구현.

미 구현사항 >
1. 자동완성이 바로 추가 되지 않고 300ms 지연.
2. 일치하는 단어 색깔 하이라이팅.
3. 자동완성이 노출 된 상태에서 위/아래 키로 포커스 이동.
4. 리스트가 선택 된 상태에서 엔터 키 입력 flow.

Todo List >
1. 매직넘버 제거.
2. view 중 하나를 class 말고 다른 형태의 prototype 패턴 사용.
3. 현재보다 나은 구조가 있을지 확인해보기.

dev-angelo committed:
[수정] 사용자 입력에 대한 매칭단어 추출 부분 수정

사용자 입력에 대한 매칭단어 추출 부분 수정.

Why >
불필요한 연산을 없애기 위해.

dev-angelo committed:
[수정] 매직넘버 제거

URL 을 하드코딩 하지 않고 enum 에 정의된 string 을 참조하도록 수정.

dev-angelo committed:
[수정] Font-Awesome 을 사용 할 수 있도록 CDN 삽입

Font-Awesome 의 아이콘을 사용하기 위해 CDN 삽입.

Why >
Input 우측의 돋보기 아이콘을 표시하기 위함.

dev-angelo committed:
[수정] [미션]검색자동완성 요구사항 기능 추가

[미션]검색자동완성 요구사항 기능 추가.

1. 자동완성 결과가 나타나 있는 상태에서 위/아래키로 이동하여
   특정 list element 에 포커스 효과가 나도록 기능 추가.

2. 자동완성 결과가 나타나 있는 상태에서 위/아래키를 통해
   특정 list element 가 focus 가 되어있는 상태에서
   Enter 키를 누르면 list element 의 text 가 input field 로 전달되고
   검정색 배경이 사라지도록 기능 추가.

3. 마우스를 특정 특정 list element 위로 올렸을 때 hover 가 되는
   기능 추가.

Todo >
1. 자동완성 결과가 300ms 후에 추가는 기능 추가/
2. 자동완성 list element 중 검색어와 일치하는 단어가
   하이라이트 되는 기능 추가.
3. (선택기능) 검색창이 포커스 되면 최근 검색 결과 5개 노출.

dev-angelo committed:
[수정] 자동완성 결과가 300ms 후에 나타나도록 수정

자동완성 결과가 300ms 후에 나타나도록 수정.

Todo >
1. 자동완성 list element 중 검색어와 일치하는 단어가
   하이라이트 되는 기능 추가.
2. (선택기능) 검색창이 포커스 되면 최근 검색 결과 5개 노출.

dev-angelo committed:
[수정] 오타 수정 및 data request 를 위한 url 수정

오타 수정 및 data request 를 위한 url 수정.

dev-angelo committed:
[수정] Server 의 프로토콜을 http 에서 https 로 수정

Server 의 프로토콜을 http 에서 https 로 수정

Why >
Github page 에서 http 프로토콜을 사용 한 req/res 가 허용되지 않음.

dev-angelo committed:
[수정] 변수들의 이름 변경

변수들의 이름 변경.

Why >
Private 변수들이기 때문.

dev-angelo committed:
[수정] Search Status 를 나타내기 위한 Enum 추가

Search Status 를 나타내기 위한 Enum 추가.

crongro commented:
> View 는 사용자에게 보여지는 내용들을 위한 계층인데,
> render 함수 외에도 사용자에게 화면이 변경되는 내용들을 설정하는 DOM 조작
> 
> 1. controller 가 서버로부터 받은 response 로 model 을 설정 하고,
> 2. controller 는 model 로부터에서 suggestion list 가 변경되었다는 것을 전달 받아
> 3. 그 내용을 view 로 전달하여
> 4. view 에서는 appendChild 등의 함수를 통해 사용자에게 보여질 자동완성 리스트 구성 등
> 
> 도 View 에서의 역할이라고 생각 할 수 있을까요?

view가 데이터 조작과 변경을 할 수도 있긴하지만
이미 설명하신대로
controller  같은 중간 단계가 view에서 보여줄 데이터를 가공하는 것도 좋죠.
개인적으로는 view가 순수해지면 view를 재사용할 수 있어서 좀더 좋은거 같아요.

crongro commented on Search/SearchAutoCompletionView.js:
```diff
@@ -0,0 +1,74 @@
+        this.searchSuggestion = null;
+    }
+
+    render() {
```
render 라는 이름 좋아요.

crongro commented on Search/SearchAutoCompletionView.js:
```diff
@@ -0,0 +1,74 @@
+        this.searchSuggestion = document.querySelector(".searchSuggestion");
+    }
+
+    onNotifySuggestionChanged(suggestion) {
```
controller에서 접근하는 경우 on으로 구별지어서 이름지으신 거 괜찮네요.

crongro commented on Search/SearchAutoCompletionView.js:
```diff
@@ -0,0 +1,74 @@
+            this._clearList();
+            this._appendList(suggestion);
+            this._changeVisibility("visible");
+            this._changeHeight(this.searchSuggestion.childElementCount * parseInt(getComputedStyle(this.searchSuggestion.firstElementChild, null).getPropertyValue("height")));
```
이 가로로 뻗어 있는 코드를 빨리 ...해결해주세요 ㅎ

crongro commented on Search/SearchAutoCompletionView.js:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    _clearList() {
+        while(this.searchSuggestion.hasChildNodes()) {
```
반복문 잘 썼음!

crongro commented on Search/SearchBackgroundView.js:
```diff
@@ -0,0 +1,39 @@
+class SearchBackgroundView {
```
backgroundview 라는 이름이 구조를 의미하니까, 맥락을 파악하기 어렵네요.

crongro commented on Search/SearchController.js:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    _appendEventHandler() {
+        const search = document.querySelector('.search');
```
searchNode라고만 해도 좀더 나을 듯.

crongro commented on Search/SearchController.js:
```diff
@@ -0,0 +1,110 @@
+        const search = document.querySelector('.search');
+
+        search.addEventListener('input', event => this._handleInputEvent(event));
+        search.addEventListener('click', event => this._handleClickEvent(event));
```
이벤트 등록 3개를 세줄로 표현한게 간결하고 맥락파악도 잘 되고 좋네요.

crongro commented on Search/SearchController.js:
```diff
@@ -0,0 +1,110 @@
+        search.addEventListener('keydown', event => this._handleKeyEvent(event));
+    }
+
+    _appendBind() {
```
private 를 _ 붙여서 표현한거 좋긴해요. 자주 사용하기도 하고요.
앞으로는 이런속성이 추가된다고 하니, 살펴보세요.
https://github.com/tc39/proposal-private-methods

crongro commented on Search/SearchController.js:
```diff
@@ -0,0 +1,110 @@
+            event.preventDefault();
+        }
+        else if (event.key === 'Enter') {
+            const focusedElement = (document.querySelector(".searchSuggestion").querySelectorAll("li"))[this._model.getCurrentIndex()];
```
여기도 가로가 너무 긴 코드..좀 줄이면 좋겠습니다. 
변수를 활용해보세요.

crongro commented on Search/SearchController.js:
```diff
@@ -0,0 +1,110 @@
+        .then(suggestionData => {
+            setTimeout(() => {
+                this._onExtractedWordsReceived(suggestionData)
+            }, 300);
```
300 하드코딩 보다는 미리 이름을 지어주고 사용하는 것이 좋을 듯.

crongro commented on Search/SearchController.js:
```diff
@@ -0,0 +1,110 @@
+            mode: 'cors',
+            cache: 'no-cache',
+            headers: {
+                'Content-Type': 'application/json',
```
json형태로 보내는거군요. 좋아요.

crongro commented on main.js:
```diff
@@ -1,6 +1,11 @@
-    const url = 'http://127.0.0.1:8080';
+    const url = 'https://ello.dlinkddns.com:8080';
+
+    const searchController = new SearchController(new SearchModel(), new SearchView());
```
객체를 생성해서 이렇게 주입시켜주는 거 좋네요.

crongro reviewed (changes requested):
컨트롤러 역할을 둔 게 인상적이네요.
좋은 시도고, 역할도 괜찮아보여요. 특히 view와 모델의 의존성을 줄인점이 잘한거죠. 
🥇 

기능이 많아지면,  컨트롤러가 매우 커질 수도 있다는 점을 염두해야겠고요, 
view에서 일어나는 event 등록을 파악하기 어렵다는 점도 단점일 수 있죠.

어떤 패턴이던 장단점이 있는데, 스스로 이부분을 판단해보는 연습이 필요합니다. 

시간되면, 
- debounce 기능을 도전해보세요 (타이핑시간 간격이 일정시간 멈추어 있으면 서버로 전송하도록)
- 캐시해서 서버로 안가도록 해보세요.(메모리에 캐시)

일단 request changes 상태로 둡니다. push하세요!

dev-angelo committed:
[수정] 불필요한 주석 제거

html 파일에서 불 필요한 주석 제거.

dev-angelo committed:
[수정] .gitignore 파일에 node_module 및 pem 파일을 무시하도록 수정

1. 기존 node_module 은 root 이하 디렉토리에 있으면 무시가 되지 않았음.
2. pem 파일은 ssl 인증을 위한 것이므로 무시 처리.

dev-angelo committed:
[수정] 리팩토링

리팩토링

Why >

1. 이전 구조는 Event 가 발생하는 지점을 파악하기 힘들었음.
2. Controller 에서는 발생 할 수 있는 이벤트를 처리하기 위한 인터페이스만
   제공하면 된다고 판단. (Event 발생은 View)
3. 또, 이전 구조로는 최근 검색어를 표현하기 힘든 점이 있었음.

dev-angelo committed:
[수정] suggestion list 를 캐싱 할 수 있도록 수정

사용자 입력에 대한 suggestion list 를 캐싱 할 수 있도록 수정.

dev-angelo commented:
> 컨트롤러 역할을 둔 게 인상적이네요.
> 좋은 시도고, 역할도 괜찮아보여요. 특히 view와 모델의 의존성을 줄인점이 잘한거죠.
> 🥇
> 
> 기능이 많아지면, 컨트롤러가 매우 커질 수도 있다는 점을 염두해야겠고요,
> view에서 일어나는 event 등록을 파악하기 어렵다는 점도 단점일 수 있죠.
> 
> 어떤 패턴이던 장단점이 있는데, 스스로 이부분을 판단해보는 연습이 필요합니다.
> 
> 시간되면,
> 
> * debounce 기능을 도전해보세요 (타이핑시간 간격이 일정시간 멈추어 있으면 서버로 전송하도록)
> * 캐시해서 서버로 안가도록 해보세요.(메모리에 캐시)
> 
> 일단 request changes 상태로 둡니다. push하세요!

질문이 있습니다~

* debounce 관련
debounce 관련 요구사항을 말씀 해 주셨는데, 사용자 입력중 dim 처리 된 영역 클릭 등으로 인하여 dim background & 추천단어 가 사라지고 input field 가 focus 를 잃어버린 후에 사용자가 다시 input field 를 클릭 하여 focus 상태가 됐을때를 의미하는 것인가요?

* 캐싱 관련
미션과는 관련이 없지만, 실제 서비스에서는 서버의 특정 알고리즘에 따라서 같은 단어를 
입력하더라고 한달전의 추천어와 오늘의 추천어가 다를 수 있는데 
이때의 캐싱은 이번 미션과의 캐싱과는 다르다고 봐도 괜찮을까요?
(이번 미션에서는 Server 에서의 추천 단어가 바뀌지 않는다는것을 가정하기 때문.)
```
실제 서비스에서는 캐싱유무와 별개로 사용자 입력이 발생하면 
1. Client 는 서버에 사용자 입력에 대한 추천단어 전달.
2. 서버는 사용자 입력에 대한 메타데이터? 를 Client 에게 전달
3. Client 는 메타데이터를 키값으로 하는 데이터가 캐싱 되어 있는지 확인.
3-1. 캐싱 되어 있다면 suggestion 은 요청하지 않음
3-2. 캐싱이 되어 있지 않다면 suggestion 요청
4. 3-2 의 경우 서버는 suggestion 전달

와 비슷한 과정을 거치는것이 맞는것인지 궁금합니다.
```

crongro commented on Search/SearchAutoCompletionView.js:
```diff
@@ -1,62 +1,64 @@
-        }
+    _appendEventHandler() {
+        this._searchDimmedBackground.addEventListener('click', event => this._onDimmedBackgroundClicked(event));
+        this._searchSuggestion.addEventListener('click', event => this._onSuggestionClicked(event));
```
this때문에 화살표 함수로 감싼거죠? 그렇다면 괜찮네요! bind를 사용한 해법도 알아두시고요.

crongro commented on Search/SearchController.js:
```diff
@@ -4,106 +4,112 @@ const SEARCH_STATUS = SEARCH_ENUM.SEARCH_STATUS;
+            suggestionClickHandler: this._handleSuggestionClick.bind(this),
+            searchButtonClickHandler: this._handleSearchButtonClick.bind(this),
+            arrowUpPressHandler: this._handleArrowUpPress.bind(this),
+            arrowDownPressHandler: this._handleArrowDownPress.bind(this),
```
bind를 다 해주었는데, 정말 다 필요한 것인지는 한번 확인하시고 사용하시고요.

dev-angelo replied:
넵! 이 부분 확인 해 보도록 하겠습니다

crongro commented on Search/SearchController.js:
```diff
@@ -4,106 +4,112 @@ const SEARCH_STATUS = SEARCH_ENUM.SEARCH_STATUS;
-                this._model.setCurrentText(event.target.value)
-                this._fetchExtractedWords(event.target.value);
-            }
+        if (cachedData) {
```
else문 없애기.

```
if(cachedData) { 
  setTimeout();
  return
}

fetch().then().then();
```

crongro commented on Search/SearchController.js:
```diff
@@ -4,106 +4,112 @@ const SEARCH_STATUS = SEARCH_ENUM.SEARCH_STATUS;
+    }
+
+    _handleSuggestionEnter(event) {
+        const focusedElement = (document.querySelector(".searchSuggestion").querySelectorAll("li"))[this._model.getCurrentIndex()];
```
너무 긴 표현이에요. 중간 값을 변수로 저장해주세요~

dev-angelo replied:
리팩토링을 진행을 먼저 해서 리뷰 받은 부분에 대해서 다시 pr 하도록 하겠습니다~!

crongro reviewed (approved):
수정을 잘하고 계시네요.
(_)를 변수에 붙이는 건 필수는 아닌 거 같아요. ㄱ ㅐ발자마다 의견이 다른부분이고요.

crongro commented:
> * debounce 관련 요구사항을 말씀 해 주셨는데, 사용자 입력중 dim 처리 된 영역 클릭 등으로 인하여 dim background & 추천단어 가 사라지고 input field 가 focus 를 잃어버린 후에 사용자가 다시 input field 를 클릭 하여 focus 상태가 됐을때를 의미하는 것인가요?

아니요. 
예를들어 입력중에는 서버로 요청이 안가는 것이고요.  예를들어서 500ms동안 입력이 없다고 판단되면 서버로 요청을 보내는거죠.

crongro commented:
> * 미션과는 관련이 없지만, 실제 서비스에서는 서버의 특정 알고리즘에 따라서 같은 단어를
>   입력하더라고 한달전의 추천어와 오늘의 추천어가 다를 수 있는데
>   이때의 캐싱은 이번 미션과의 캐싱과는 다르다고 봐도 괜찮을까요?
>   (이번 미션에서는 Server 에서의 추천 단어가 바뀌지 않는다는것을 가정하기 때문.)

네 맞아요. 원래는 실시간 변경이 될 수 있는건 캐시하면 안되겠죠 ^^
